contributors/design-proposals/storage/csi-migration.md

```diff
@@ -661,6 +695,93 @@ with minimal disruption to users would be:
 And at this point users doing their own deployment and not installing the GCE PD
 CSI driver encounter an error.
 
+## Disabling in-tree plugin code
+
+Before we can stop compiling and prepare to remove the code associated with migrated
+in-tree plugins, we need to make sure all persistent volume operations involving
+in-tree plugins continue to function in a backward compatible way when the
+in-tree plugin code paths are disabled. When CSIMigrationInTreeOff feature flag
+is enabled, we will disable in-tree plugin code by skipping probing of migrated
+in-tree plugins in Kubelet's ProbeVolumePlugins or Kube Controller Manager's
+ProbeAttachableVolumePlugins, ProbeExpandableVolumePlugins and
+ProbeControllerVolumePlugins. All functions in kubernetes code base that depend
+on probed plugins need to be audited and modified to handle nil plugins returned
+as detailed below.
+
+### Enhancements in ProbeVolumePlugins
+
+A new file pkg/volume/util/csimigration/plugins.go will have the following functions:
+```
+func ProbePluginsWithCSIMigrationDisabled(plugins []volume.VolumePlugin) []volume.VolumePlugin
+func ProbeAttachablePluginsWithCSIMigrationDisabled(plugins []volume.VolumePlugin) []volume.VolumePlugin
+```
+They will [1] check plugin specific migration feature flags as well as [2] the overall
+CSIMigrationInTreeOff feature flag to determine whether ProbeVolumePlugins
+of a legacy in-tree plugin will get invoked. Kubelet's ProbeVolumePlugins and Kube
+Controller Manager's ProbeAttachableVolumePlugins, ProbeExpandableVolumePlugins and
+ProbeControllerVolumePlugins will call the above instead of directly probing in-tree plugins
+that will be migrated.
+
+Once CSIMigrationInTreeOff feature flag and the plugin specific migration
+flags get enabled by default, the build tag `nolegacyproviders` can be enabled on
+pkg/volume/util/csimigration/plugins.go so that no in-tree plugins get probed.
+
+### Detection of migration status for a plugin
+
+Code paths that need to check migration status of a plugin, for example,
+in provisionClaimOperationExternal, findDeletablePlugin, etc. will need to
+depend on a common utility function:
+```
+func IsCSIMigrationEnabledForPluginByName(pluginName string) bool
+```
+that will be introduced in pkg/volume/csi_migration.go
+
+Note that a per-plugin member function of the form IsMigratedToCSI cannot be used
+since [1] a plugin object for in-tree plugins will typically be nil and [2] the
+code implementing IsMigratedToCSI will be removed as part of disabling plugin code.
+
+### Handling of nil Plugin Object returned by FindPluginBySpec/FindPluginByName
+
+Once an in-tree plugin is no longer probed, all the VolumePluginMgr
+functions of the form Find*PluginBySpec/Find*PluginByName will return a nil plugin
```

Comment:
won't they also return an error? This should just be an error path saying "plugin not found" or whatever it is currently

Reply:
Correct. I will change the above sentence accordingly.